src/orchestrator/streaming/aggregator.rs

```diff
@@ -0,0 +1,484 @@
+struct GenerationActor {
+    rx: mpsc::Receiver<GenerationActorMessage>,
+    generations: Vec<ClassifiedGeneratedTextStreamResult>,
+}
+
+impl GenerationActor {
+    pub fn new(rx: mpsc::Receiver<GenerationActorMessage>) -> Self {
+        let generations = Vec::new();
+        Self { rx, generations }
+    }
+
+    async fn run(&mut self) {
+        while let Some(msg) = self.rx.recv().await {
+            self.handle(msg);
+        }
+    }
+
+    fn handle(&mut self, msg: GenerationActorMessage) {
+        match msg {
+            GenerationActorMessage::Put(generation) => self.generations.push(generation),
+            GenerationActorMessage::Get { index, response_tx } => {
+                let generation = self.generations.get(index).cloned();
+                let _ = response_tx.send(generation);
+            }
+            GenerationActorMessage::GetRange {
+                start,
+                end,
+                response_tx,
+            } => {
+                let generations = self.generations[start..=end].to_vec();
+                let _ = response_tx.send(generations);
+            }
+            GenerationActorMessage::Length { response_tx } => {
+                let _ = response_tx.send(self.generations.len());
+            }
+        }
+    }
+}
+
+/// [`GenerationActor`] handle.
+struct GenerationActorHandle {
+    tx: mpsc::Sender<GenerationActorMessage>,
+}
+
+impl GenerationActorHandle {
+    pub fn new() -> Self {
+        let (tx, rx) = mpsc::channel(32);
+        let mut actor = GenerationActor::new(rx);
+        tokio::spawn(async move { actor.run().await });
+        Self { tx }
+    }
+
+    pub async fn put(&self, generation: ClassifiedGeneratedTextStreamResult) {
+        let msg = GenerationActorMessage::Put(generation);
+        let _ = self.tx.send(msg).await;
+    }
+
+    pub async fn get(&self, index: usize) -> Option<ClassifiedGeneratedTextStreamResult> {
+        let (response_tx, response_rx) = oneshot::channel();
+        let msg = GenerationActorMessage::Get { index, response_tx };
+        let _ = self.tx.send(msg).await;
+        response_rx.await.unwrap()
+    }
+
+    pub async fn get_range(
+        &self,
+        start: usize,
+        end: usize,
+    ) -> Vec<ClassifiedGeneratedTextStreamResult> {
+        let (response_tx, response_rx) = oneshot::channel();
+        let msg = GenerationActorMessage::GetRange {
+            start,
+            end,
+            response_tx,
+        };
+        let _ = self.tx.send(msg).await;
+        response_rx.await.unwrap()
+    }
+
+    pub async fn len(&self) -> usize {
+        let (response_tx, response_rx) = oneshot::channel();
+        let msg = GenerationActorMessage::Length { response_tx };
+        let _ = self.tx.send(msg).await;
+        response_rx.await.unwrap()
+    }
+}
+
+#[derive(Debug, Clone)]
+struct Tracker {
+    state: Vec<Option<(Span, Detections)>>,
```

Comment:
One reason we were using BTreeMap previously was that it is automatically sorted and provides easy access. So the thought was that while the insertion would be slightly costlier, it would be cheaper to access the elements later on.  

So we can keep the spans as unit for indexing (which we are doing here, in new version as well), but have them available in order of span. Which would make it easier to check things like `x` span is returned or not, etc. Which is possible even in this implementation by `find_by_span_start`, but might be tricky to do things like `first_key_value`, which I remember was also benefit for one of the edge cases 🤔

Reply:
I'm not seeing any sort requirement or benefit of using `BTreeMap` for how this is being used. `Vec` is simple and fast for the current `Tracker` use case.